CDLUC3/dmsp_backend_prototype
# Initial template schema, MariaDB for docker and general cleanup of mocks

## Description

Initial schema, model, resolver (queries only), database tables and seed data for a Template. This PR is for #14 and based off of initial definition in #33

All SQL logic lives in the Models now, with Resolvers simply making calls to the model to perform those queries. In the future, the Resolvers will perform authorization checks and business logic (e.g. check that the user has permission, generate a new version of the template, save the new version, notify stakeholders, etc.)

Additions:
- Replaced the old ApolloServerContext mock with a new one that is more accurate and usable
- Added `MySqlModel` abstract class that manages common properties and functions
- Added `Collaborator` abstract class and `TemplateCollaborator` class (models, schema, and resolvers)
- Added `Template` class (model, schema, and resolver)
- Added `VersionedTemplate` class (model, schema, and resolver)
- Added `TemplateService` that performs `generateVersion` and `clone`

Updates:
- Move Affiliation API calls from the Resolver to the Model 
- Updated `ContributorRole` and `User` models, resolvers and schema to be like the new models
- Updated the `src/context.ts` to have a callable `buildContext()` function so that it is usable through the normal Apollo server context as well as within our Express endpoints for `login` and `register`
- Changed some `private` properties in classes to `protected` because they were giving me grief in Jest mocks. (Unable to mock a private property/function)
- Removed the `Model` suffix on some of the model classes so that names are consistent
- Removed old `src/config/AwsConfig.ts`. We weren't using it and when we do need to access AWS resources, aws-sdk v3 is different
- Removed all of the old Apollo server mocks (except Affiliation ... left that as an example in case we want to revisit) since we could not get that working and all of the tables have been implemented.
- Updated all other `data-migration/*.sql` files so that the tables all have the standard: `created`, `modified`, `id`, `createdById` and `modifiedById` fields.
- Reordered a few of the `data-migration/` SQL files so that they run in a correct order so the DB can be built from scratch

Other bits:
- Fixed the ESLint config and added it as a husky pre-commit task
- Updated the README with accurate info
- removed old MySQL config from `.env.example` (
- Added MariaDB to docker-compose
- Added script to `data-migrations/` that drops all of the tables in the local DB

### Testing
- Update your local `.env` and remove any references to MySQL
- startup the container: `docker-compose up`
- Nuke your existing DB: `docker-compose exec apollo bash ./data-migrations/nuke-db`
- Rebuild the DB with new tables and seed data: `docker-compose exec apollo bash ./data-migrations/process.sh`
- Visit the GraphQL explorer `http://localhost:4000/graphql` and check out the new Queries
